lib/rust/ensogl/pack/src/assets.rs

```diff
@@ -0,0 +1,362 @@
+use crate::Paths;
+use enso_prelude::anyhow;
+use ide_ci::prelude::*;
+use ide_ci::programs::shaderc::Glslc;
+use ide_ci::programs::shaderc::SpirvOpt;
+use ide_ci::programs::spirv_cross::SpirvCross;
+use std::hash::Hasher;
+
+
+
+// =============
+// === Build ===
+// =============
+
+/// Build the dynamic assets.
+pub async fn build(paths: &Paths) -> Result<()> {
+    info!("Building dynamic assets.");
+    let sources = survey_asset_sources(paths)?;
+    let assets = update_assets(paths, &sources).await?;
+    let manifest = serde_json::to_string(&assets)?;
+    std::fs::write(&paths.target.ensogl_pack.dist.dynamic_assets.manifest, manifest)?;
+    gc_assets(paths, &assets)?;
+    Ok(())
+}
+
+
+
+// ===============
+// === Builder ===
+// ===============
+
+/// Identifies an asset type, which determines how it is built.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, PartialOrd, Ord, Serialize)]
+#[serde(rename_all = "lowercase")]
+enum Builder {
+    Font,
+    Shader,
+}
+
+impl Builder {
+    fn dir_name<'a>(self) -> &'a str {
+        self.into()
+    }
+
+    async fn build_asset(
+        self,
+        input_dir: &Path,
+        input_files: &[String],
+        output_dir: &Path,
+        tmp: &Path,
+    ) -> Result<()> {
+        match self {
+            Builder::Font => build_font(input_dir, input_files, output_dir).await,
+            Builder::Shader => build_shader(input_dir, input_files, output_dir, tmp).await,
+        }
+    }
+}
+impl TryFrom<&str> for Builder {
+    type Error = anyhow::Error;
+    fn try_from(value: &str) -> std::result::Result<Self, Self::Error> {
+        match value {
+            "font" => Ok(Builder::Font),
+            "shader" => Ok(Builder::Shader),
+            other => Err(anyhow!("Unknown builder: {other:?}")),
+        }
+    }
+}
+impl From<Builder> for &'static str {
+    fn from(value: Builder) -> Self {
+        match value {
+            Builder::Font => "font",
+            Builder::Shader => "shader",
+        }
+    }
+}
+
+
+
+// ====================
+// === Build Inputs ===
+// ====================
+
+/// The inputs to a builder.
+struct AssetSources {
+    asset_key:   String,
+    input_files: Vec<String>,
+    inputs_hash: u64,
+}
+
+impl AssetSources {
+    /// The output directory name for the asset.
+    fn dir_name(&self) -> String {
+        let key = &self.asset_key;
+        let hash = self.inputs_hash;
+        format!("{key}-{hash:x}")
+    }
+}
+
+
+
+// =====================
+// === Build Outputs ===
+// =====================
+
+/// The outputs of a builder.
+#[derive(Serialize)]
+struct Asset {
+    dir:   String,
+    files: Vec<String>,
+}
+
+/// The outputs of all builders.
+type AssetManifest = BTreeMap<Builder, BTreeMap<String, Asset>>;
+
+
+
+// ================
+// === Building ===
+// ================
+
+/// Scan the sources found in the asset sources directory.
```

Comment:
1. The docs tell that this functions scans the sources. But this function is pretty complex and returns something. Docs should describe what is the result of the scan.
2. This and the following 3 functions are pretty complex with really minimal docs. I'm missing some bigger docs describing what all of they do from high perspective.